x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/mergers/rule_source/calculate_rule_source_from_asset.test.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RuleResponse } from '../../../../../../../../common/api/detection_engine';
+import type { PrebuiltRuleAsset } from '../../../../../prebuilt_rules';
+import { calculateRuleSourceFromAsset } from './calculate_rule_source_from_asset';
+
+const buildTestRule = (overrides?: Partial<RuleResponse>) => {
+  return {
+    rule_id: 'rule_id',
+    version: 1,
+    ...overrides,
+  } as RuleResponse;
+};
+
+const buildTestRuleAsset = (overrides?: Partial<PrebuiltRuleAsset>) => {
+  return {
+    rule_id: 'rule_id',
+    version: 1,
+    source_updated_at: '2024-05-01',
+    ...overrides,
+  } as PrebuiltRuleAsset;
+};
```

Comment:
You can use here:

```
import { getRulesSchemaMock } from '../../../../../../../../common/api/detection_engine/model/rule_schema/mocks';
import { getPrebuiltRuleMock } from '../../../../../prebuilt_rules/model/rule_assets/prebuilt_rule_asset.mock';
```
instead of this function, otherwise Zod validation is failing in the tests.

Reply:
Spent some considerable time with this test suite to make sure all our assumptions were correct, so I ended up rewriting the test suite. I'm copying it here so it isn't wasted, feel free to use this code:
```ts
import { calculateRuleSourceFromAsset } from './calculate_rule_source_from_asset';
import { getRulesSchemaMock } from '../../../../../../../../common/api/detection_engine/model/rule_schema/mocks';
import { getPrebuiltRuleMock } from '../../../../../prebuilt_rules/model/rule_assets/prebuilt_rule_asset.mock';

describe('calculateRuleSourceFromAsset', () => {
  it('calculates as internal if no asset is found', () => {
    const result = calculateRuleSourceFromAsset({
      rule: getRulesSchemaMock(),
      assetWithMatchingVersion: undefined,
      ruleIdExists: false,
    });

    expect(result).toEqual({
      type: 'internal',
    });
  });

  it('calculates as customized external type if an asset is found, but no asset with a matching version', () => {
    const ruleToImport = getRulesSchemaMock();
    const result = calculateRuleSourceFromAsset({
      rule: ruleToImport,
      assetWithMatchingVersion: undefined,
      ruleIdExists: true,
    });

    expect(result).toEqual({
      type: 'external',
      is_customized: true,
    });
  });

  describe('matching rule_id and version is found', () => {
    it('calculates as customized external type if the imported rule has all fields unchanged from the asset', () => {
      const ruleToImport = getRulesSchemaMock();
      const result = calculateRuleSourceFromAsset({
        rule: getRulesSchemaMock(), // version 1
        assetWithMatchingVersion: getPrebuiltRuleMock({
          ...ruleToImport,
          version: 1, // version 1 (same version as imported rule)
          // no other overwrites -> no differences
        }),
        ruleIdExists: true,
      });

      expect(result).toEqual({
        type: 'external',
        is_customized: false,
      });
    });

    it('calculates as non-customized external type the imported rule has fields which differ from the asset', () => {
      const ruleToImport = getRulesSchemaMock();
      const result = calculateRuleSourceFromAsset({
        rule: getRulesSchemaMock(), // version 1
        assetWithMatchingVersion: getPrebuiltRuleMock({
          ...ruleToImport,
          version: 1, // version 1 (same version as imported rule)
          name: 'Customized name', // mock a customization
        }),
        ruleIdExists: true,
      });

      expect(result).toEqual({
        type: 'external',
        is_customized: true,
      });
    });
  });
});
```
I did have to edit the `[getPrebuiltRuleMock](https://github.com/rylnd/kibana/blob/b88dac31a80894848889cc46f0ac11ffd675695d/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.mock.ts)` method to get allow to overwrite its mocked fields:
```ts
import { PrebuiltRuleAsset } from './prebuilt_rule_asset';

export const getPrebuiltRuleMock = (rewrites?: Partial<PrebuiltRuleAsset>): PrebuiltRuleAsset => {
  const parsedFields = rewrites ? PrebuiltRuleAsset.parse(rewrites) : {};

  return {
    description: 'some description',
    name: 'Query with a rule id',
    query: 'user.name: root or user.name: admin',
    severity: 'high',
    type: 'query',
    risk_score: 55,
    language: 'kuery',
    rule_id: 'rule-1',
    version: 1,
    ...parsedFields,
  };
};
```